server/util/retry/retry.go

```diff
@@ -180,3 +186,53 @@ func (r *Retry) MaxAttempts() int {
 func (r *Retry) MaxTotalDelay() time.Duration {
 	return r.maxTime
 }
+
+// Do executes the given function with a retry loop.
+//
+// The caller can indicate that an error should not be retried by wrapping it
+// using NonRetryableError(err).
+func Do[T any](ctx context.Context, opts *Options, fn func(ctx context.Context) (T, error)) (T, error) {
+	r := New(ctx, opts)
+	var lastError error
+	for r.Next() {
+		rsp, err := fn(ctx)
+		if err != nil {
+			lastError = err
+			continue
+		}
+		if lastError != nil && !opts.DontLogFailedAttempts {
+			name := opts.Name
+			if name != "" {
+				name += " "
+			}
+			log.CtxWarningf(ctx, "%sattempt failed, but succeeded on retry: %s", name, err)
```

Comment:
i'm a little fuzzy on this bit -- is it gonna log something like `GetActionResultattempt failed`?

I think it would be useful to have a String() method on the Retry that is safe to call and shows `attempt $x of $y`

Reply:
This will log a single warning, but only if the function eventually succeeds. 

If all retries are exhausted, we will return an error so there's no need to log anything. It's up to the caller to decide if they want to log in that case. Since we don't know whether the operation will eventually succeed, I postpone logging the failed attempt until we return a non-error response.